tests/test.py

```diff
@@ -9859,6 +9861,311 @@ def testTransactionRollback(self):
+                WHERE id = $1 OR id = $2
+            """
+
+            stmt = await conn.prepare(select_two_query)
+
+            with self.assertRaises(asyncpg.exceptions.SyntaxOrAccessError) as ex:
+                await stmt.fetch(corrupted_data['id'], data['id'])
+
+            self.assertEqual(ex.exception.message,
+                             'encoding error in column "value_bytes"')
+
+            # Insert and select new data using the same connection to be sure
+            # it doesn't close or get out of sync
+            data = {
+                'id': get_random_id(),
+                'value_bytes': random_bytes(),
+            }
+            await conn.execute(insert_query, data['id'], data['value_bytes'])
+            row = await conn.fetchrow(select_query, data['id'])
+            self.assertEqual(data['value_bytes'], row['value_bytes'])
+
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(test())
+
+    def testTransactionPreparedRollback(self):
+        """
+        Test that connection is not closed in case of "encoding error" with
+        prepared statement and a driver can do rollback in transaction after
+        that.
+        """
+        async def test():
+            self.encryptor_table.create(bind=self.engine_raw, checkfirst=True)
+            # Insert data that will trigger decryption error
+            corrupted_data = {
+                'id': get_random_id(),
+                'value_bytes': random_bytes(),
+            }
+            self.engine_raw.execute(
+                self.encryptor_table.insert(), corrupted_data)
+            data = {
+                'id': get_random_id(),
+                'value_bytes': random_bytes(),
+            }
+
+            conn = await self.executor.connect()
+
+            insert_query = """
+                INSERT INTO test_proper_db_flushing_on_error(id, value_bytes)
+                VALUES ($1, $2)
+            """
+            select_query = """
+                SELECT value_bytes
+                FROM test_proper_db_flushing_on_error
+                WHERE id = $1
+            """
+
+            with self.assertRaises(asyncpg.exceptions.SyntaxOrAccessError) as ex:
+                async with conn.transaction():
+                    await conn.execute(insert_query, data['id'], data['value_bytes'])
+
+                    row = await conn.fetchrow(select_query, data['id'])
+                    self.assertEqual(data['value_bytes'], row['value_bytes'])
+
+                    stmt = await conn.prepare(select_query)
+                    # Expect encoding error
+                    await stmt.fetch(corrupted_data['id'])
+
+            self.assertEqual(ex.exception.message,
+                             'encoding error in column "value_bytes"')
+            row = await conn.fetchrow(select_query, data['id'])
+            self.assertEqual(row, None)
+
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(test())
```

Comment:
I fiound that python has high-level helper [asyncio.run](https://docs.python.org/3/library/asyncio-task.html#asyncio.run) for that. Lets start to use it?) Works since python3.7.

Reply:
I wanted to use that, because I use it in small scripts myself. But I've read from the docs:
> This function cannot be called when another asyncio event loop is running in the same thread.
I don't know whether our test run in parallel, but if they do, this can cause a problem (or I just don't understand this statement correctly)

Reply:
then lets create new event loop for that?) use 'new_event_loop` instead of `get_event_loop`

Reply:
Okay, but I don't quite understand the reasons. Could you explain why?

Reply:
it will help run code in parallel if need. every thread or test case will have own loop and it will safe to stop loop on failure and don't touch other tests that can rely on common loop in current thread